### PR TITLE
CRM-17303 Fix to ensure you cannot schedule Mailings for in the past

### PR DESCRIPTION
**Pull request by seamuslee001**
- [CRM-17303: CiviMails can be scheduled for dates in past](https://issues.civicrm.org/jira/browse/CRM-17303)

### CODE REVIEW

**Comment by seamuslee001** [comment]
The test failures are definitely an issue and would appreciate some thoughts on it from the core team. @colemanw can you take a look. The CRM Alert was an idea from one of our users wanting a user prompt when they entered a bad date not just greying out of the submit box

**Comment by totten** [comment]
Maybe I'm missing something, but `crmUiDatepicker` doesn't seem quite right -- that directive should be usable for many purposes (e.g. scheduling mailings or editing past activities). Prohibiting past dates is only appropriate to certain fields. Seems like either (a) it should be an optional behavior or (b) it should be implemented at a different level (eg `crmMailing/BlockSchedule*`).

**Comment by colemanw** [comment]
@seamuslee001 as totten said I think this is the right logic in the wrong place.
@totten would you recommend field-level validation or form-level (e.g. when you hit the send button)?

**Comment by seamuslee001** [comment]
@colemanw @totten one thing is that for some reason in crmMailing/BlockSchedule i can't use the require ngModal so am struggling to work out how to input that as i would assume i need that to set the $setValidity

**Comment by seamuslee001** [comment]
@colemanw @totten I have shifted it to the radiodate.js file which is in crmMailing folder and seems to be more correct sort of place to do this validation. I tried in crmMailing/BlockSchedule but angular would constantly give me errors like https://docs.angularjs.org/error/$compile/ctreq?p0=ngModel&p1=crmMailingBlockSchedule when i tried to do a require on ngModel and i couldn't do $scope.SetValidity either. I found that this was the only other place i could reliably do what i was looking for. If either of you have suggestions as to how i could get it done in crmMailing/BlockSchedule.js I'm more than willing to give it a shot

**Comment by seamuslee001** [comment]
Jenkins test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: ang/crmMailing/RadioDate.js [hunk 1]
```diff
@@ -48,6 +48,26 @@
...
+              var submittedDate = $(this).val();
+              if (month.length < 2) month = '0' + month;
+              if (day.length < 2) day = '0' + day;
+              date = [year, month, day].join('-');
+              time = [hours, minutes, "00"].join(':');
+              currentDate = date + ' ' + time;
+              ngModel.$setValidity('dateTimeInThePast', !($(this).val().length && submittedDate < currentDate));
+              if ($(this).val().length && submittedDate < currentDate) {
+                crmUiAlert({
+                  text: ts('The Scheulded date and time is in the past'),
```
Comment (davecivicrm) [inline ang/crmMailing/RadioDate.js]: @seamuslee001 Typo in alert message - 'Scheulded' should be 'scheduled'.

Comment (seamuslee001) [inline ang/crmMailing/RadioDate.js]: @davecivicrm Thanks for picking it up dave have now fixed it

**Comment by seamuslee001** [comment]
@colemanw I think this should be right now, if you want to take another look.

**Comment by colemanw** [comment]
this looks good to me. what do you think @totten ?

**Comment by seamuslee001** [comment]
@totten Tim you fine with this PR?

**Comment by totten** [comment]
Yup, shifting it to crmMailingRadioDate makes sense to me.

**Comment by seamuslee001** [comment]
@colemanw i think we are good to go, I squished some commits down using git rebase

**Comment by colemanw** [comment]
Excellent. Thanks for seeing this through.

**Comment by colemanw** [comment]
@seamuslee001 it looks like the unit tests are failing since merging this:

> ALERT: 'Error
> The scheduled date and time is in the past'

**Comment by seamuslee001** [comment]
@colemanw I don't think that is indicating it is failing at all, all that is indicating is that the dates are in the past. which they are by the unit test. so its just that because the dates / times are in the past the unit test is triggering the alert get activated

**Comment by colemanw** [comment]
Probably the test needs updating to use future dates.

**Comment by seamuslee001** [comment]
in my pr i did update the tests and added a new test which should be setting the date in the future, if you look at reports like https://test.civicrm.org/job/CiviCRM-Core-PR/6043/testReport/ it seems that its only been your css branch that was failing for some reason. but that one seemed to pass fine and was against master. No JS unit test errors

**Comment by colemanw** [comment]
I'm referring to https://test.civicrm.org/job/CiviCRM-Core-Matrix/CIVIVER=master,label=ubuntu1204/1696/console

Our [main test suite](https://test.civicrm.org/job/CiviCRM-Core-Matrix/CIVIVER=master,label=ubuntu1204/) looks to be failing now.

**Comment by seamuslee001** [comment]
@colemanw I'm not sure it is because looking at that report it says "+ karma start --browsers PhantomJS --single-run --reporters dots,junit
[32mINFO [karma]: [39mKarma v0.12.37 server started at http://localhost:9876/
[32mINFO [launcher]: [39mStarting browser PhantomJS
[32mINFO [PhantomJS 1.9.8 (Linux 0.0.0)]: [39mConnected on socket 7TK_WLeahaT4Ku9itPcp with id 26678775
...................
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
.
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
ALERT: 'Error
The scheduled date and time is in the past'
.........
PhantomJS 1.9.8 (Linux 0.0.0): Executed 29 of 29 SUCCESS (0.535 secs / 0.808 secs)" Note it says all 29 passed. so even tho it is putting out the alert the tests are passing

**Comment by seamuslee001** [comment]
@colemanw looking through the log i see 

not ok 216 - Failure: CRM_Contribute_ActionMapping_ByTypeTest::testDefault with data set #2 ('2015-02-02 00:00:00', 'addAliceDues addBobDonation scheduleForDonation startWeekBefore repeatTwoWeeksAfter useHelloFirstName', array(array('2015-01-26 00:00:00', array('bob@example.org'), '/Hello, Bob.*via subject/'), array('2015-02-02 00:00:00', array('bob@example.org'), '/Hello, Bob.*via subject/'), array('2015-02-09 00:00:00', array('bob@example.org'), '/Hello, Bob.*via subject/'), array('2015-02-16 00:00:00', array('bob@example.org'), '/Hello, Bob.*via subject/')))
  ---
  message: 'Failure in api call for job send_reminder:  DB Error: unknown error'
  severity: fail
  data:
    got: 1
    expected: 0
  ...
PHP Fatal error:  Call to a member function reset() on a non-object in /srv/buildkit/build/build-2/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php on line 508
+ EXITCODE=1
+ ./scripts/phpunit --tap --log-junit=/var/lib/jenkins/workspace/CiviCRM-Core-Matrix/CIVIVER/master/label/ubuntu1204/junit/api_v3_AllTests.xml api_v3_AllTests

which i think is the actual failure nothing in relation to this PR

**Comment by colemanw** [comment]
Ok thanks for looking.

**Pull request merged**